// @sentry/react-router/build/cjs/index.server.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const node = require('@sentry/node');
const sdk = require('./server/sdk.js');
const wrapSentryHandleRequest = require('./server/wrapSentryHandleRequest.js');
const createSentryHandleRequest = require('./server/createSentryHandleRequest.js');
const wrapServerAction = require('./server/wrapServerAction.js');
const wrapServerLoader = require('./server/wrapServerLoader.js');
const createSentryHandleError = require('./server/createSentryHandleError.js');
const getMetaTagTransformer = require('./server/getMetaTagTransformer.js');
const createServerInstrumentation = require('./server/createServerInstrumentation.js');
const plugin = require('./vite/plugin.js');
const handleOnBuildEnd = require('./vite/buildEnd/handleOnBuildEnd.js');
const makeConfigInjectorPlugin = require('./vite/makeConfigInjectorPlugin.js');
const serverGlobals = require('./server/serverGlobals.js');



exports.init = sdk.init;
exports.sentryHandleRequest = wrapSentryHandleRequest.sentryHandleRequest;
exports.wrapSentryHandleRequest = wrapSentryHandleRequest.wrapSentryHandleRequest;
exports.createSentryHandleRequest = createSentryHandleRequest.createSentryHandleRequest;
exports.wrapServerAction = wrapServerAction.wrapServerAction;
exports.wrapServerLoader = wrapServerLoader.wrapServerLoader;
exports.createSentryHandleError = createSentryHandleError.createSentryHandleError;
exports.getMetaTagTransformer = getMetaTagTransformer.getMetaTagTransformer;
exports.createSentryServerInstrumentation = createServerInstrumentation.createSentryServerInstrumentation;
exports.sentryReactRouter = plugin.sentryReactRouter;
exports.sentryOnBuildEnd = handleOnBuildEnd.sentryOnBuildEnd;
exports.makeConfigInjectorPlugin = makeConfigInjectorPlugin.makeConfigInjectorPlugin;
exports.isInstrumentationApiUsed = serverGlobals.isInstrumentationApiUsed;
Object.prototype.hasOwnProperty.call(node, '__proto__') &&
	!Object.prototype.hasOwnProperty.call(exports, '__proto__') &&
	Object.defineProperty(exports, '__proto__', {
		enumerable: true,
		value: node['__proto__']
	});

Object.keys(node).forEach(k => {
	if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) exports[k] = node[k];
});
