// @sentry/react-router/build/cjs/cloudflare/index.js
Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const core = require('@sentry/core');
const browser = require('@sentry/browser');
const sdk = require('../client/sdk.js');
const tracingIntegration = require('../client/tracingIntegration.js');
const react = require('@sentry/react');
const createClientInstrumentation = require('../client/createClientInstrumentation.js');
const wrapSentryHandleRequest = require('../server/wrapSentryHandleRequest.js');

/**
 * Injects Sentry trace meta tags into the HTML response by transforming the ReadableStream.
 * This enables distributed tracing by adding trace context to the HTML document head.
 * @param body - ReadableStream containing the HTML response body to modify
 * @returns A new ReadableStream with Sentry trace meta tags injected into the head section
 */
function injectTraceMetaTags(body) {
  const headClosingTag = '</head>';

  const reader = body.getReader();
  const stream = new ReadableStream({
    async pull(controller) {
      const { done, value } = await reader.read();

      if (done) {
        controller.close();
        return;
      }

      const encoder = new TextEncoder();
      const html = value instanceof Uint8Array ? new TextDecoder().decode(value) : String(value);

      if (html.includes(headClosingTag)) {
        const modifiedHtml = html.replace(headClosingTag, `${core.getTraceMetaTags()}${headClosingTag}`);

        controller.enqueue(encoder.encode(modifiedHtml));
        return;
      }

      controller.enqueue(encoder.encode(html));
    },
  });

  return stream;
}

exports.init = sdk.init;
exports.reactRouterTracingIntegration = tracingIntegration.reactRouterTracingIntegration;
exports.ErrorBoundary = react.ErrorBoundary;
exports.Profiler = react.Profiler;
exports.captureReactException = react.captureReactException;
exports.reactErrorHandler = react.reactErrorHandler;
exports.useProfiler = react.useProfiler;
exports.withErrorBoundary = react.withErrorBoundary;
exports.withProfiler = react.withProfiler;
exports.createSentryClientInstrumentation = createClientInstrumentation.createSentryClientInstrumentation;
exports.isClientInstrumentationApiUsed = createClientInstrumentation.isClientInstrumentationApiUsed;
exports.isNavigateHookInvoked = createClientInstrumentation.isNavigateHookInvoked;
exports.wrapSentryHandleRequest = wrapSentryHandleRequest.wrapSentryHandleRequest;
exports.injectTraceMetaTags = injectTraceMetaTags;
Object.prototype.hasOwnProperty.call(browser, '__proto__') &&
  !Object.prototype.hasOwnProperty.call(exports, '__proto__') &&
  Object.defineProperty(exports, '__proto__', {
    enumerable: true,
    value: browser['__proto__']
  });

Object.keys(browser).forEach(k => {
  if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) exports[k] = browser[k];
});
