// @sentry/react-router/build/esm/server/instrumentation/reactRouter.js
import { InstrumentationBase, InstrumentationNodeModuleDefinition } from '@opentelemetry/instrumentation';
import { SEMATTRS_HTTP_TARGET } from '@opentelemetry/semantic-conventions';
import { SDK_VERSION, debug, getActiveSpan, getRootSpan, spanToJSON, updateSpanName, SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN, SEMANTIC_ATTRIBUTE_SENTRY_SOURCE, startSpan, SEMANTIC_ATTRIBUTE_SENTRY_OP } from '@sentry/core';
import { DEBUG_BUILD } from '../../common/debug-build.js';
import { isInstrumentationApiUsed } from '../serverGlobals.js';
import { isDataRequest, getOpName, getSpanName } from './util.js';

const supportedVersions = ['>=7.0.0'];
const COMPONENT = 'react-router';

/**
 * Instrumentation for React Router's server request handler.
 * This patches the requestHandler function to add Sentry performance monitoring for data loaders.
 */
class ReactRouterInstrumentation extends InstrumentationBase {
   constructor(config = {}) {
    super('ReactRouterInstrumentation', SDK_VERSION, config);
  }

  /**
   * Initializes the instrumentation by defining the React Router server modules to be patched.
   */
  // eslint-disable-next-line @typescript-eslint/naming-convention
   init() {
    const reactRouterServerModule = new InstrumentationNodeModuleDefinition(
      COMPONENT,
      supportedVersions,
      (moduleExports) => {
        return this._createPatchedModuleProxy(moduleExports);
      },
      (_moduleExports) => {
        // nothing to unwrap here
        return _moduleExports;
      },
    );

    return reactRouterServerModule;
  }

  /**
   * Creates a proxy around the React Router module exports that patches the createRequestHandler function.
   * This allows us to wrap the request handler to add performance monitoring for data loaders and actions.
   */
   _createPatchedModuleProxy(moduleExports) {
    return new Proxy(moduleExports, {
      get(target, prop, receiver) {
        if (prop === 'createRequestHandler') {
          const original = target[prop];
          return function sentryWrappedCreateRequestHandler( ...args) {
            const originalRequestHandler = original.apply(this, args);

            return async function sentryWrappedRequestHandler(request, initialContext) {
              let url;
              try {
                url = new URL(request.url);
              } catch {
                return originalRequestHandler(request, initialContext);
              }

              // We currently just want to trace loaders and actions
              if (!isDataRequest(url.pathname)) {
                return originalRequestHandler(request, initialContext);
              }

              // Skip OTEL instrumentation if instrumentation API is being used
              // as it handles loader/action spans itself
              if (isInstrumentationApiUsed()) {
                DEBUG_BUILD && debug.log('Skipping OTEL loader/action instrumentation - using instrumentation API');
                return originalRequestHandler(request, initialContext);
              }

              const activeSpan = getActiveSpan();
              const rootSpan = activeSpan && getRootSpan(activeSpan);

              if (!rootSpan) {
                DEBUG_BUILD && debug.log('No active root span found, skipping tracing for data request');
                return originalRequestHandler(request, initialContext);
              }

              // We cannot rely on the regular span name inferral here, as the express instrumentation sets `*` as the route
              // So we force this to be a more sensible name here
              // TODO: try to set derived parameterized route from build here (args[0])
              const spanData = spanToJSON(rootSpan);
              // eslint-disable-next-line deprecation/deprecation
              const target = spanData.data[SEMATTRS_HTTP_TARGET] || url.pathname;
              updateSpanName(rootSpan, `${request.method} ${target}`);
              rootSpan.setAttributes({
                [SEMANTIC_ATTRIBUTE_SENTRY_SOURCE]: 'url',
                [SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: 'auto.http.react_router.server',
              });

              return startSpan(
                {
                  name: getSpanName(url.pathname, request.method),
                  attributes: {
                    [SEMANTIC_ATTRIBUTE_SENTRY_ORIGIN]: 'auto.http.react_router.server',
                    [SEMANTIC_ATTRIBUTE_SENTRY_OP]: getOpName(url.pathname, request.method),
                  },
                },
                () => {
                  return originalRequestHandler(request, initialContext);
                },
              );
            };
          };
        }
        return Reflect.get(target, prop, receiver);
      },
    });
  }
}

export { ReactRouterInstrumentation };
